### PR TITLE
Feature/file size formatting

### PR DESCRIPTION
**Pull request by mattheworiordan**
Closes https://discourse.metabase.com/t/format-or-style-field-in-bytes/25384

### Description

This PR adds a new "File size" number formatting style that automatically converts byte values to human-readable units (B, KB, MB, GB, TB, PB).

The implementation supports both binary (1024-based) and decimal (1000-based) unit systems, with automatic scaling to the most appropriate unit. Columns containing byte-related keywords in their names (like file_size, bandwidth, bytes) are automatically detected and default to file size formatting.

Key features:
  - New "File size" option in the Style dropdown for numeric columns
  - Binary units (KiB, MiB, GiB, TiB, PiB) and decimal units (KB, MB, GB, TB, PB)
  - Configurable unit display location (column header vs individual cells)
  - Consistent formatting across frontend and backend
  - Works with tables, charts, and exports

### How to verify

Describe the steps to verify that the changes are working as expected.

1. New question → Sample Dataset → Create a Native Query with test data:
```sql
SELECT
    'small.txt' as filename,
    1024 as file_size
  UNION ALL SELECT 'medium.jpg', 5242880
  UNION ALL SELECT 'large.zip', 1073741824
  UNION ALL SELECT 'huge.iso', 5368709120
```

2. Visualize as Table → Click gear icon on file_size column

3. Under "Style", select "File size" (should auto-detect based on column name)

4. Test the "Unit system" toggle:
    - Binary: Shows 1 KiB, 5 MiB, 1 GiB, 5 GiB
    - Decimal: Shows 1.02 KB, 5.24 MB, 1.07 GB, 5.37 GB

5. For table visualizations, test "Where to display the unit":
    - "In every table cell": Shows units with each value
    - "In the column heading": Shows "File Size (Bytes)" in header, numbers only in cells

6. Create a bar chart with the same data to verify Y-axis formatting

7. Download results as CSV to confirm formatting is preserved

### Demo

MO screenshot 2025-08-15 at 19 43 11

- Table with file size formatting applied
- Settings panel with new "File size" option
- Binary vs decimal units comparison

MO screenshot 2025-08-15 at 19 52 43

- Chart with formatted Y-axis values]

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

**Comment by dpsutton** [comment]
nice! goals are a bit weird but that's not anything new here
image

Looking at the dashboard and the "static-viz":
image
image
We need to do the formatting in the static viz bundle as well.

**Comment by dpsutton** [comment]
the backend tests are failing for me:
image

```clojure
Ran 6 tests containing 215 assertions.
18 failures, 0 errors.
{:test 6, :pass 197, :fail 18, :error 0, :type :summary}
```

Many of these stem from an edge case that the minimum threshold of unit is `1` but the minimum value is `0`.

```clojure
(map #(numbers/format-number % {:number-style "filesize"}) [0 0.1 0.4 0.6 0.9 2 5])
("0.0 B" "0.1 B" "0.4 B" "0.6 B" "0.9 B" "2 B" "5 B")
```

The expectations in the tests are that these will have no fractional value since fractional bytes don't exist. This comes from this line: `[threshold unit] (first (filter #(>= abs-value (first %)) units))`. i haven't looked at the other few test failures when that happens.

File: frontend/src/metabase/visualizations/shared/settings/column.ts [hunk 1]
```diff
@@ -13,6 +13,14 @@ export function getDefaultNumberStyle(
     return "percent";
   }
 
+  // Auto-detect file size columns
+  const columnName = column.name?.toLowerCase() || "";
+  const bytesIndicators = ["byte", "size", "bandwidth", "traffic", "data_transfer", "file_size"];
```
Comment (alxnddr) [inline frontend/src/metabase/visualizations/shared/settings/column.ts]: There has to be a proper semantic type for this similar to how it works for other columns. For instance, `isCurrency` checks for the semantic type set by the backend.

File: frontend/src/metabase/visualizations/lib/settings/column.js [hunk 2]
```diff
@@ -396,6 +445,11 @@ export const NUMBER_COLUMN_SETTINGS = {
           return getCurrencySymbol(settings["currency"]);
         }
         return getCurrency(settings["currency"], settings["currency_style"]);
+      } else if (
+        settings["number_style"] === "filesize" &&
+        settings["filesize_unit_in_header"]
+      ) {
+        return "Bytes";
```
Comment (alxnddr) [inline frontend/src/metabase/visualizations/lib/settings/column.js]: this should be localized t`Bytes`

**Comment by mattheworiordan** [comment]
@dpsutton @alxnddr thanks for the feedback. As this is my first PR on Metabase, are you expecting me to address this feedback now, or is this something you're picking up on your end?  Can you explain - I'm happy to look if needed, but just need some clarification on what's expected please. 🙏

**Comment by dragonsahead** [comment]
@mattheworiordan we're expecting a fix from your end

**Comment by mattheworiordan** [comment]
Thanks, I believe the changes I have made in https://github.com/metabase/metabase/pull/62235/commits/8d995a02ca6863936913b1e971df343815130ff6 addresses all feedback.

I cannot rerun CI so will need you to confirm if tests pass like they do locally for me.

Some screenshots:

MO screenshot 2025-08-31 at 08 03 59


MO screenshot 2025-08-31 at 08 05 39


MO screenshot 2025-08-31 at 08 05 47

File: docs/file-size-formatting.md [hunk 3]
```diff
@@ -0,0 +1,228 @@
+# File Size Formatting
```
Comment (alxnddr) [inline docs/file-size-formatting.md]: User-facing documentation related to formatting is stored in `docs/data-modeling/formatting.md`, please extend it instead. This file looks ai-generated anyway as it refers to other files that do not exist so feel free to just remove it

File: src/metabase/types/core.cljc [hunk 4]
```diff
@@ -133,6 +133,13 @@
 (derive :type/Duration :Semantic/*)
 (derive :type/Duration :type/Number)
 
+;; :type/FileSize -- represents data sizes in bytes (file sizes, bandwidth, data transfer, storage capacity)
+(derive :type/FileSize :Semantic/*)
```
Comment (alxnddr) [inline src/metabase/types/core.cljc]: We probably don't need so many separate semantic types for data quantity. Let's see what @metabase/core-backend-querying @metabase/core-frontend-querying think

File: frontend/src/metabase-lib/v1/types/utils/isa.js [hunk 5]
```diff
@@ -182,6 +182,9 @@ export const isCurrency = (field) =>
 export const isPercentage = (field) =>
   field && isa(field.semantic_type, TYPE.Percentage);
 
+export const isFileSize = (field) =>
```
Comment (alxnddr) [inline frontend/src/metabase-lib/v1/types/utils/isa.js]: As it describes not only file sizes but bandwith, storage capacity, it would be great to have a more generic name for the semantic type, number style type, and related utils such as `TYPE.DataMeasure`, `isDataMeasure`, `dataMeasure`

File: frontend/src/metabase/lib/formatting/numbers.tsx [hunk 6]
```diff
@@ -215,6 +244,11 @@ function formatNumberCompact(
         ...COMPACT_CURRENCY_OPTIONS,
       });
 
+      // nf should never be null for currency formatting, but TypeScript needs the check
+      if (!nf) {
+        return String(value);
```
Comment (alxnddr) [inline frontend/src/metabase/lib/formatting/numbers.tsx:249]: Maybe better to throw for `_formatNumberCompact` fallback, but this should not happen anyway

File: frontend/test/metabase/lib/formatting/filesize-integration.unit.spec.js [hunk 7]
```diff
@@ -0,0 +1,161 @@
+// Integration tests for file size formatting feature
```
Comment (alxnddr) [inline frontend/test/metabase/lib/formatting/filesize-integration.unit.spec.js]: These are not integration tests, they test `formatNumber` in isolation. The tests for some reason create some unnecessary data structures, then call formatNumber on some properties of it and perform assertions on these properties which practically is not different from tests in `frontend/src/metabase/lib/formatting/numbers.unit.spec.ts`

File: frontend/test/metabase/lib/formatting/filesize-integration.unit.spec.js [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+      const columns = [
+        { name: "id" },
+        { name: "file_size" },
+        { name: "bandwidth" },
+        { name: "bytes_transferred" },
+        { name: "revenue" },
+      ];
+
+      // This simulates what getDefaultNumberStyle would do
+      const shouldUseFileSize = (column) => {
```
Comment (alxnddr) [inline frontend/test/metabase/lib/formatting/filesize-integration.unit.spec.js]: this test tests function that is defined in the test

File: frontend/test/metabase/lib/formatting/filesize-integration.unit.spec.js
Comment (alxnddr) [inline frontend/test/metabase/lib/formatting/filesize-integration.unit.spec.js:1]: This entire file seems very llm-hallucinated

**Review by alxnddr — CHANGES_REQUESTED** [review]
Separator style setting does not change the separator style with DataMeasure formatting:
Screenshot 2025-09-04 at 10 07 48 PM

Also, compact formatting removes data measure formatting style

File: src/metabase/types/core.cljc (hunk 4 above)
Comment (mattheworiordan) [reply src/metabase/types/core.cljc]: Ok, can you suggest what you want here as this is your code base and your customers. I think the four types are meaningfully distinct.

File: frontend/test/metabase/lib/formatting/filesize-integration.unit.spec.js
Comment (mattheworiordan) [reply frontend/test/metabase/lib/formatting/filesize-integration.unit.spec.js:1]: Sure, this is Claude assisted, which I assume is expected given https://github.com/metabase/metabase/blob/master/CLAUDE.md.

Can you be more specific about what issue you see here so that I can address it?

File: docs/file-size-formatting.md (hunk 3 above)
Comment (mattheworiordan) [reply docs/file-size-formatting.md]: Good spot, this was unnecessary. I will remove it and check if here is anything in the link you provided that needs updating.

File: frontend/src/metabase-lib/v1/types/utils/isa.js (hunk 5 above)
Comment (mattheworiordan) [reply frontend/src/metabase-lib/v1/types/utils/isa.js]: Thanks for the suggestion, I did struggle to lock in one name that works. DataMeasure works for me, I will address that.

File: frontend/src/metabase/lib/formatting/numbers.tsx (hunk 6 above)
Comment (mattheworiordan) [reply frontend/src/metabase/lib/formatting/numbers.tsx:249]: OK, will do

File: frontend/test/metabase/lib/formatting/filesize-integration.unit.spec.js (hunk 7 above)
Comment (mattheworiordan) [reply frontend/test/metabase/lib/formatting/filesize-integration.unit.spec.js]: Ok, so you happy for me to drop these set of tests completely?

File: frontend/test/metabase/lib/formatting/filesize-integration.unit.spec.js (hunk 8 above)
Comment (mattheworiordan) [reply frontend/test/metabase/lib/formatting/filesize-integration.unit.spec.js]: Yup, good spot. Will correct

**Comment by mattheworiordan** [comment]
I have addressed all feedback I can in c0269d1. I hope this gets this PR over the line, but shout if you still need something else done.

File: frontend/test/metabase/lib/formatting/filesize-integration.unit.spec.js (hunk 7 above)
Comment (mattheworiordan) [reply frontend/test/metabase/lib/formatting/filesize-integration.unit.spec.js]: I have now droppe them completely

**Comment by thebiglabasky** [comment]
Thanks for the proposal!

I wonder about the term "File size" since when applied to something like bandwidth it may read odd. I guess files likely being the most common it makes sense, otherwise I'd suggest a more generic name for the type like "Byte size"?

I'll give it a spin but I think it's a nice addition.

**Comment by thebiglabasky** [comment]
After checking internally with @jeff-bruemmer we'd ideally prefer "Data size" as a name so this applies better to various places using bytes 🙏🏻

**Comment by mattheworiordan** [comment]
> we'd ideally prefer "Data size" as a name so this applies better to various places using bytes 🙏🏻

Ok, done in 71f1885

WDYT?

**Comment by mattheworiordan** [comment]
Hi. It's been 3 weeks since I addressed all your feedback and this PR now appears to be stale for no shared reason. Can you please tell me what the blocker is?

**Review by ranquild — COMMENTED** [review]
@mattheworiordan please leave only number formatting changes and new formatting options and remove other changes, e.g. new semantic types. We are not sure about the semantic type additions and not ready to merge them in, but new formatting options are fine. Please ping us if/when it's ready for a final review. Thanks!

**Comment by mattheworiordan** [comment]
@ranquild done -> 23ee45c

File: frontend/src/metabase/visualizations/shared/settings/column.ts [hunk 9]
```diff
@@ -1,6 +1,13 @@
 import { isCurrency, isPercentage } from "metabase-lib/v1/types/utils/isa";
 import type { ColumnSettings, DatasetColumn } from "metabase-types/api";
 
+function hasDataSizeName(column: DatasetColumn): boolean {
```
Comment (alxnddr) [inline frontend/src/metabase/visualizations/shared/settings/column.ts]: We should not hardcode column type detection like this on the frontend, this is done via semantic type system like below in `getDefaultNumberStyle` function via utils like `isCurrency` or `isPercentage` that check for semantic types. Since based on @ranquild input we don't want to add a new semantic type here, we need to remove this auto-detection code because it is a wrong place to do it

File: frontend/src/metabase/lib/formatting/numbers.tsx (hunk 6 above)
Comment (alxnddr) [reply frontend/src/metabase/lib/formatting/numbers.tsx:249]: Would still be nice to add error throwing

File: frontend/src/metabase/lib/formatting/numbers.tsx [hunk 10]
```diff
@@ -180,11 +200,20 @@ export function formatChangeWithSign(
...
+export function numberFormatterForOptions(
+  options: FormatNumberOptions,
+): Intl.NumberFormat | null {
   options = {
     ...getDefaultNumberOptions(options),
     ...options,
   };
+
+  // datameasure is a custom format, not supported by Intl.NumberFormat
+  // Return null for datameasure - formatNumber will handle it specially
```
Comment (alxnddr) [inline frontend/src/metabase/lib/formatting/numbers.tsx]: There are a few places where comments or code still use `datameasure` or `filesize` instead of `datasize`

File: frontend/src/metabase/visualizations/lib/settings/column.js [hunk 11]
```diff
@@ -316,6 +324,49 @@ export const NUMBER_COLUMN_SETTINGS = {
     },
     readDependencies: ["number_style"],
   },
+  datasize_unit_system: {
```
Comment (alxnddr) [inline frontend/src/metabase/visualizations/lib/settings/column.js:327]: Please add one end-to-end test to `e2e/test/scenarios/visualizations-tabular/table.cy.spec.js` that verifies the formatting is applied to the table content, and `datasize_unit_in_header` setting work

**Review by alxnddr — COMMENTED** [review]
Great progress! I left some final comments from my side

File: frontend/src/metabase/visualizations/shared/settings/column.ts (hunk 9 above)
Comment (mattheworiordan) [reply frontend/src/metabase/visualizations/shared/settings/column.ts]: Ok, so just to be clear, are you therefore saying there should now be no automatic detection of columns that are clearly a data size type and only rely on a user configuring that, or have I misunderstood your comment here?

File: frontend/src/metabase/lib/formatting/numbers.tsx (hunk 10 above)
Comment (mattheworiordan) [reply frontend/src/metabase/lib/formatting/numbers.tsx]: Oops, will correct.

File: frontend/src/metabase/visualizations/shared/settings/column.ts (hunk 9 above)
Comment (mattheworiordan) [reply frontend/src/metabase/visualizations/shared/settings/column.ts]: @alxnddr can I get a response please so that I can finish up this PR? I do need some clarity on this point please. 
Thanks!

Comment (alxnddr) [reply frontend/src/metabase/visualizations/shared/settings/column.ts]: @mattheworiordan yes, that is correct. Since we're not adding a proper semantic type here which must be used for auto detection and going with just a visual setting then there should not be this frontend-side formatting detection code. When users want this formatting, they will have to manually specify it in column settings

**Comment by mattheworiordan** [comment]
Hi @alxnddr, I've addressed all the outstanding feedback:

Changes in latest commit:

1. Removed frontend auto-detection - Deleted hasDataSizeName() and its usage in getDefaultNumberStyle(). Data size formatting now only applies when manually selected by users, as requested.
2. Fixed naming consistency - Updated all comments referencing "filesize" or "datameasure" to use "datasize" consistently.
3. Added E2E tests and all passing locally - Added 3 tests to table.cy.spec.js covering:
    - Manual data size formatting application (1024 → 1 KiB)
    - Binary vs decimal unit system toggle
    - datasize_unit_in_header setting (unit in header vs cells)
    
Would you be able to take another look when you have a chance? Happy to make any additional changes if needed. 

Really hoping we can get this merged now 🙏 

_Note: Just seen there is a merge conflict now. Can you approve this in principle before I rebase and destroy the commit history and comments_

**Comment by mattheworiordan** [comment]
@alexyarosh @alxnddr any update?

**Comment by mattheworiordan** [comment]
Hi team - this PR has now been through 7 iterations over 5 months, with me addressing every piece of feedback:

* Added semantic types (per @alxnddr's request)
* Removed semantic types (per @ranquild's contradicting request)
* Renamed from FileSize → DataMeasure → DataSize
* Fixed separator style, compact formatting bugs
* Removed frontend auto-detection
* Added E2E tests

The last substantive feedback was addressed on Nov 28 - nearly 2 months ago. I've been waiting for review since then.

Can someone please clarify what's blocking this small yet obviously needed feature? Happy to rebase once we have approval in principle. @alxnddr @ranquild @johnswanson @eliotmetabase

**Comment by alxnddr** [comment]
Hey @mattheworiordan 👋  I'm very sorry for the long review cycle, we could've done better job about it. We don't have many external contributions therefore our workflow is not build around them but here specifically my reactions were way slow.

The feature is very useful but other items on our roadmap took our attention. As for the pull request feedback, we have certain feature and code quality standards for merging. I have just tested it manually again and I noticed that [this issue](https://github.com/metabase/metabase/pull/62235#issuecomment-3229547954) can still be reproduced:

**Test Dashboard (formatting is correct)**
Screenshot 2026-01-19 at 2 22 36 PM

**Exported card (missing formatting)**
Screenshot 2026-01-19 at 2 23 11 PM

**Dashboard subscriptions (missing formatting)**

Screenshot 2026-01-19 at 2 23 41 PM

Screenshot 2026-01-19 at 2 23 31 PM


I'll check with the team if we can prioritize this feature for you and implement it internally with all cases covered and get back to you.

**Comment by ghostframe** [comment]
We've decided not to implement this feature at this time

**Comment by mattheworiordan** [comment]
> We've decided not to implement this feature at this time

Why not? How do you propose we provide reporting to our customers for data metrics now given there is no workaround/no plugins/no flexibility here?  Quite honestly the only option I think I have is a run a fork (terrible idea for so many reason) or move to another product (feeling that way given it's taken 5 months and multiple revisions and changes in direction to not add a feature that is [clearly needed by others](https://discourse.metabase.com/t/format-or-style-field-in-bytes/25384) and objectively a reasonable ask). 

What are you recommending now as a paying customer of Metabase?

**Comment by likeshumidity** [comment]
@mattheworiordan one potential workaround:
Use a Custom Column with a CASE statement like below:

```
case(
  [Total] < 1024,
  concat(round([Total]), "B"),
  [Total] < 1048576,
  concat(round([Total] / 1024 * 10) / 10, "KB"),
  [Total] < 1073741824,
  concat(round([Total] / 1048576 * 10) / 10, "MB"),
  [Total] < 1099511627776,
  concat(round([Total] / 1073741824 * 10) / 10, "GB"),
  concat(
    round([Total] / 1099511627776 * 10) / 10,
    "TB"
  )
)
```

**Comment by mattheworiordan** [comment]
Thanks for your response @likeshumidity, however if you look at the reason this PR was created (link at the top to https://discourse.metabase.com/t/format-or-style-field-in-bytes/25384), that solution does not work for many reasons:

- Users cannot export data with appropriate fidelity or indeed really any sensible way to import into any other system because the values are not real values
- You cannot chart/graph this data
- You cannot sort on these values

... and the list goes on.

So this PR was a solution to that, and could have provided a simple base for other unit types.
